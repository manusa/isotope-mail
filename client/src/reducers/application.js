import {INITIAL_STATE} from './';
import {ActionTypes} from '../actions/action-types';

const application = (state = INITIAL_STATE.application, action = {}) => {
  switch (action.type) {
    case ActionTypes.APPLICATION_BE_REQUEST:
      return {...state, activeRequests: state.activeRequests + 1};
    case ActionTypes.APPLICATION_BE_REQUEST_COMPLETED:
      return {...state, activeRequests: state.activeRequests > 0 ? state.activeRequests - 1 : 0};
    case ActionTypes.APPLICATION_USER_CREDENTIALS_SET:
      return {
        ...state, user: {
          ...state.user,
          id: action.payload.userId, hash: action.payload.hash, credentials: action.payload.credentials
        }
      };
    case ActionTypes.APPLICATION_FOLDER_SELECT:
      return {
        ...state,
        selectedFolderId: action.payload.folderId
      };
    case ActionTypes.APPLICATION_FOLDER_RENAME: {
      return {
        ...state,
        renameFolderId: action.payload ? action.payload.folderId : null
      };
    }
    case ActionTypes.APPLICATION_FOLDER_RENAME_OK: {
      const newState = {...state};
      if (state.selectedFolderId === action.payload.oldFolderId) {
        newState.selectedFolderId = action.payload.newFolderId;
      }
      return newState;
    }
    case ActionTypes.APPLICATION_MESSAGE_SELECT:
      return {...state, selectedMessage: {...action.payload}};
    case ActionTypes.APPLICATION_MESSAGE_REFRESH: {
      const newState = {...state};
      const message = action.payload.message;
      // Store in application.downloadedMessages
      newState.downloadedMessages[message.messageId] = message;
      // Update selected message if is currently selected
      if (state.selectedFolderId === action.payload.folder.folderId
        && state.selectedMessage.uid === message.uid) {
        newState.selectedMessage = {...message};
      }
      return newState;
    }
    case ActionTypes.APPLICATION_MESSAGE_REFRESH_BE_REQUEST:
      return {...state, refreshMessageActiveRequests: state.refreshMessageActiveRequests + 1};
    case ActionTypes.APPLICATION_MESSAGE_REFRESH_BE_REQUEST_COMPLETED:
      return {...state, refreshMessageActiveRequests:
          state.refreshMessageActiveRequests > 0 ? state.refreshMessageActiveRequests - 1 : 0};
    case ActionTypes.APPLICATION_MESSAGE_PRE_DOWNLOAD: {
      const newState = {...state};
      const messages = action.payload.messages;
      // Store in application.downloadedMessages
      messages.forEach(message => {
        newState.downloadedMessages[message.messageId] = message;
      });
      return newState;
    }
    case ActionTypes.APPLICATION_MESSAGE_REPLACE_IMAGE: {
      const newState = {...state};
      const folder = action.payload.folder;
      const message = action.payload.message;
      const attachment = action.payload.attachment;
      const contentId = attachment.contentId.replace(/[<>]/g, '');
      const regex = new RegExp(`cid:${contentId}`, 'g'); // Multiple occurrence -> ReplaceAll
      if (contentId.length > 0) {
        const objectUrl = URL.createObjectURL(action.payload.blob);
        // Store in application.downloadedMessages
        newState.downloadedMessages[message.messageId].content =
          newState.downloadedMessages[message.messageId].content.replace(regex, objectUrl);
        // Update selected message if applicable
        if (newState.selectedFolderId === folder.folderId
          && newState.selectedMessage.uid === message.uid) {
          const parsedMessage = newState.selectedMessage.content.replace(regex, objectUrl);
          newState.selectedMessage = {...newState.selectedMessage, content: parsedMessage};
        }
      }
      return newState;
    }
    case ActionTypes.APPLICATION_ERROR_SET:
      const errorsSetState = {...state};
      errorsSetState.errors = {...state.errors};
      errorsSetState.errors[action.payload.type] = action.payload.value;
      return errorsSetState;
    case ActionTypes.APPLICATION_MESSAGE_EDIT: {
      const newState = {...state};
      newState.newMessage = action.payload ? {...action.payload} : null;
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_SEND: {
      const newState = {...state};
      newState.outbox = {message: action.payload, progress: 0, sent: false};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_UPDATE_PROGRESS: {
      const newState = {...state};
      newState.outbox = {...newState.outbox, progress: action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_SET_SENT: {
      const newState = {...state};
      newState.outbox = {...newState.outbox, sent: action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_SET_ERROR: {
      const newState = {...state};
      newState.outbox = {...newState.outbox, error: action.payload};
      return newState;
    }
    case ActionTypes.APPLICATION_OUTBOX_MESSAGE_PROCESSED: {
      const newState = {...state};
      newState.outbox = null;
      return newState;
    }
    default:
      return state;
  }
};

export default application;
