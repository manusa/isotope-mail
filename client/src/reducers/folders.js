import {initialState} from './';
import {ActionTypes} from '../actions/action-types';

const folders = (state = initialState.folders, action) => {
  switch (action.type) {
    case ActionTypes.ADD_FOLDER:
      // return {...state, folders: [...state.folders, action.payload]}
      return [...state, action.payload];
    default:
      return state;
  }
}

export default folders;
