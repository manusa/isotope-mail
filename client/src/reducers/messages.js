import {initialState} from './';
import {ActionTypes} from '../actions/action-types';

const messages = (state = initialState.messages, action) => {
  switch (action.type) {
    case ActionTypes.MESSAGES_BE_REQUEST:
      return {...state, activeRequests: state.activeRequests + 1};
    case ActionTypes.MESSAGES_BE_REQUEST_COMPLETED:
      return {...state, activeRequests: state.activeRequests > 0 ? state.activeRequests - 1 : 0};
    case ActionTypes.MESSAGES_UPDATE_CACHE:
      const newState = {...state};
      newState.cache[action.payload.folder.folderId] =
        new Map(action.payload.messages.map(m => [m.uid, m]));
      return newState;
    case ActionTypes.ADD_MESSAGE:
      return {...state, items: [...state.items, action.payload]};
    default:
      return state;
  }
};

export default messages;
