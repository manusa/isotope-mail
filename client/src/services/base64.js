// https://stackoverflow.com/a/30106551/999375
// https://tools.ietf.org/html/rfc4648#page-6 - Standard Base64 Alphabet
// https://tools.ietf.org/html/rfc4648#page-8 - Url Safe  Base64 Alphabet

export function unicodeUrlBtoa(string) {
  return btoa(encodeURIComponent(string).replace(/%([0-9A-F]{2})/g,
    (match, p1) => String.fromCharCode(`0x${p1}`)
  )).replace(/\+/g, '-').replace(/\//g, '_');
}

export function unicodeUrlAtob(string) {
  const replacedString = string.replace(/-/g, '+').replace(/_/g, '/');
  return decodeURIComponent(atob(replacedString)
    .split('')
    // eslint-disable-next-line prefer-template
    .map(c => '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2))
    .join(''));
}

