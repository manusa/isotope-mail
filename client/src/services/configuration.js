import {URLS} from './url';
import {toJson} from './fetch';

/**
 * @module services/configuration
 */
/**
 * @typedef {Object} Configuration
 * @property {Array} _links - HATEOAS links for application
 */

const TABLET_BREAKPOINT_PX = 600;

/**
 * Fetches configuration from BE
 * @returns {Promise<Configuration>}
 */
export async function fetchConfiguration() {
  const response = await fetch(URLS.CONFIGURATION);
  if (response.ok) {
    return await toJson(response);
  }
  return null;
}

/**
 * Checks if the current viewport is desktop sized
 *
 * @returns {boolean} true if current window size is a valid 'desktop' size, false otherwise
 */
export function isDesktop() {
  return Math.max(document.documentElement.clientWidth, window.innerWidth) > TABLET_BREAKPOINT_PX;
}
