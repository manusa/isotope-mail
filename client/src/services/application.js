import {setUserCredentials} from '../actions/application';

export function login(dispatch, credentials) {
  fetch('http://localhost:9010/v1/application/login', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(credentials)
  })
    .then(response => (response.json()))
    .then(json =>
      dispatch(setUserCredentials(json))
    );
}
