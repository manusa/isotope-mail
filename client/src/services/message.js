import {
  backendRequest as applicationBackendRequest,
  backendRequestCompleted as applicationBackendRequestCompleted,
  preDownloadMessages
} from '../actions/application';
import {
  backendRequest,
  backendRequestCompleted,
  deleteFromCache, lockMessages, setFolderCache,
  setSelected, unlockMessages,
  updateCache, updateCacheIfExist
} from '../actions/messages';
import {updateFolder} from '../actions/folders';
import {abortControllerWrappers, abortFetch, AuthenticationException, credentialsHeaders, toJson} from './fetch';
import {persistMessageCache} from './indexed-db';
import {notifyNewMail} from './notification';
import {FolderTypes} from './folder';

const _eventSourceWrappers = {};

function _closeEventSource(dispatch, es) {
  if (es && es.close) {
    es.close();
    if (!es.DISPATCHED) {
      dispatch(backendRequestCompleted());
      es.DISPATCHED = true;
      es.resolvePromise();
    }
  }
}

/**
 * "Aborts" the event source (SSE received stream) to reset folder message cache.
 *
 * This will cancel current message poll.
 *
 * @param dispatch
 */
export function closeResetFolderMessagesCacheEventSource(dispatch) {
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
}

/**
 *
 * @param dispatch
 * @param user
 * @param folder
 */
export async function resetFolderMessagesCache(dispatch, user, folder) {
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  if (folder && folder._links) {
    const allMessages = [];
    // Prefer EventSourcePolyfill instead of EventSource to allow sending HTTP headers in all browsers
    const es = new window.EventSourcePolyfill(folder._links.messages.href,
      {
        headers: credentialsHeaders(user.credentials)
      });
    _eventSourceWrappers.resetFolderMessagesCache = es;
    dispatch(backendRequest());
    let isFirstBatch = true;
    es.onmessage = e => {
      const messages = JSON.parse(e.data);
      allMessages.push(...messages);

      // Remove UIDs not included in batch update from store
      const originalUids = messages.map(m => m.uid);
      let maxUid = null;
      let minUid = null;
      if (originalUids.length > 0) {
        maxUid = originalUids.reduce((a, b) => Math.max(a, b));
        minUid = originalUids.reduce((a, b) => Math.min(a, b));
        const completeUidSequence = Array(maxUid - minUid + 1).fill('').map((v, i) => i + minUid);
        const uidsToRemove = completeUidSequence.filter(uid => !originalUids.includes(uid));
        if (uidsToRemove.length > 0) {
          dispatch(deleteFromCache(folder, uidsToRemove.map(uid => ({uid}))));
        }
      }

      dispatch(updateCache(folder, messages));

      // This is the First Batch -> Delete all messages with a higher UID not included in the batch
      if (isFirstBatch) {
        isFirstBatch = false;
        if (maxUid) {
          dispatch(deleteFromCache(folder, [], {from: maxUid + 1}));
        } else {
          // Clear empty folder if first batch has no uids
          dispatch(deleteFromCache(folder, [], {from: 0}));
        }
      }

      // This is the last batch -> persistMessageCache with all new gathered messages
      if (e.lastEventId === '1') {
        _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
        // Remove uids not included in batch (0-minUidInBatch)
        if (minUid) {
          dispatch(deleteFromCache(folder, [], {to: minUid - 1}));
        }
        // Manually persist newest version of message cache
        persistMessageCache(user.id, user.hash, folder, [...allMessages]);
      }
    };
    // Return a promise
    return new Promise((resolve, reject) => {
      // Store resolve function that will be called in _closeEventSource
      es.resolvePromise = resolve;
      es.onerror = ({status, message}) => {
        if (status === 401) {
          reject(new AuthenticationException(message));
        }
      };
    });
  }
  return null;
}

export function preloadMessages(dispatch, credentials, folder, messageUids) {
  if (messageUids.length === 0) {
    return;
  }
  const url = new URL(folder._links.messages.href);
  messageUids.forEach(messageUid => url.searchParams.append('id', messageUid));
  fetch(url, {
    method: 'GET',
    headers: credentialsHeaders(credentials)
  })
    .then(toJson)
    .then(messages => {
      if (folder.type === FolderTypes.INBOX && messages.some(m => m.recent === true)) {
        notifyNewMail();
      }
      dispatch(preDownloadMessages(messages));
    })
    .catch(() => {/* Ignore any preload error */});
}

export function downloadAttachment(credentials, attachment) {
  const fetch$ = fetch(attachment._links.download.href, {
    method: 'GET',
    headers: credentialsHeaders(credentials)
  }).then(response => response.blob());
  fetch$
    .then(blob => {
      if (navigator.msSaveBlob) {
        navigator.msSaveBlob(blob, attachment.fileName);
      } else {
        const url = URL.createObjectURL(blob);
        const tempLink = document.createElement('a');
        tempLink.href = url;
        tempLink.download = attachment.fileName;
        document.body.appendChild(tempLink);
        // tempLink.click();
        tempLink.dispatchEvent(new MouseEvent('click', {bubbles: true, cancelable: true, view: window}));
        document.body.removeChild(tempLink);
        URL.revokeObjectURL(url);
      }
    });
  return fetch$;
}

/**
 * Moves the provided array of messages from the original fromFolder to the target toFolder.
 *
 * This method is optimistic and will calculate fromFolder message counts before any server response is received.
 *
 * @param dispatch {!function}
 * @param credentials {!Object}
 * @param fromFolder {!Object}
 * @param toFolder {!Object}
 * @param messages {!Array.<Object>}
 */
export function moveMessages(dispatch, credentials, fromFolder, toFolder, messages) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  const fromFolderUpdated = {
    ...fromFolder
  };
  messages.forEach(m => {
    dispatch(setSelected([m], false));
    // Calculate new fromFolder values
    fromFolderUpdated.messageCount -= 1;
    fromFolderUpdated.unreadMessageCount -= (m.seen ? 0 : 1);
    fromFolderUpdated.newMessageCount -= (m.recent ? 1 : 0);
  });

  // Update state with expected response from server
  dispatch(deleteFromCache(fromFolder, messages));
  dispatch(updateFolder(fromFolderUpdated));
  dispatch(lockMessages(messages));
  fetch(fromFolder._links['message.move.bulk'].href.replace('{toFolderId}', toFolder.folderId), {
    method: 'PUT',
    headers: credentialsHeaders(credentials, {'Content-Type': 'application/json'}),
    body: JSON.stringify(messages.map(m => m.uid))
  })
    .then(toJson)
    .then(newMessages => {
      dispatch(unlockMessages(messages));
      if (Array.isArray(newMessages)) {
        dispatch(updateCache(toFolder, newMessages));
        // Update folder info with the last message (contains the most recent information)
        dispatch(updateFolder(newMessages[newMessages.length - 1].folder));
      }
    })
    .catch(() => {
      dispatch(unlockMessages(messages));
      // Rollback state from dispatched expected responses
      dispatch(updateCache(fromFolder, messages));
      dispatch(updateFolder(fromFolder));
    });
}

export function setMessagesSeen(dispatch, credentials, folder, messages, seen) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  const messagesToUpdate = [];
  // Update state with expected response from server
  const expectedUpdatedFolder = {...folder};
  messages.forEach(m => {
    // Calculate new folder values
    if (m.seen !== seen) {
      messagesToUpdate.push({...m, seen: seen});
      expectedUpdatedFolder.unreadMessageCount += seen ? -1 : +1;
    }
  });
  dispatch(updateCacheIfExist(folder, messagesToUpdate, true));
  dispatch(updateFolder(expectedUpdatedFolder));
  dispatch(lockMessages(messagesToUpdate));

  fetch(folder._links['message.seen.bulk'].href.replace('{seen}', seen.toString()), {
    method: 'PUT',
    headers: credentialsHeaders(credentials, {'Content-Type': 'application/json'}),
    body: JSON.stringify(messagesToUpdate.map(m => m.uid))
  })
    .then(response => {
      dispatch(unlockMessages(messagesToUpdate));
      if (!response.ok) {
        // Rollback state from dispatched expected responses
        dispatch(updateCacheIfExist(folder, messages));
        dispatch(updateFolder(folder));
      }
    });
}

export function setMessageFlagged(dispatch, credentials, folder, message, flagged) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  dispatch(updateCacheIfExist(folder, [{...message, flagged}], true));
  dispatch(lockMessages([message]));
  fetch(folder._links['message.flagged'].href.replace('{messageId}', message.uid), {
    method: 'PUT',
    headers: credentialsHeaders(credentials, {'Content-Type': 'application/json'}),
    body: JSON.stringify(flagged)
  })
    .then(response => {
      dispatch(unlockMessages([message]));
      if (!response.ok) {
        // Rollback state from dispatched expected responses
        dispatch(updateCacheIfExist(folder, [message]));
      }
    });
}

export async function deleteAllFolderMessages(dispatch, credentials, folder) {
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);
  dispatch(applicationBackendRequest());
  try {
    const updatedFolder = await fetch(folder._links.messages.href, {
      method: 'DELETE',
      headers: credentialsHeaders(credentials)
    }).then(toJson);
    dispatch(updateFolder(updatedFolder));
    dispatch(setFolderCache(updatedFolder, []));
    dispatch(applicationBackendRequestCompleted());
  } catch (e) {
    dispatch(applicationBackendRequestCompleted());
  }
}

export function deleteMessages(dispatch, credentials, folder, messages) {
  if (messages.length === 0) {
    return;
  }
  // Abort any operations that can affect operation result
  _closeEventSource(dispatch, _eventSourceWrappers.resetFolderMessagesCache);
  abortFetch(abortControllerWrappers.getFoldersAbortController);

  const url = new URL(folder._links.messages.href);
  const messagesToDelete = [];
  // Update state with expected response from server
  const expectedUpdatedFolder = {...folder};
  messages.forEach(message => {
    url.searchParams.append('id', message.uid);
    messagesToDelete.push({...message, deleted: true});
    expectedUpdatedFolder.deletedMessageCount++;
    expectedUpdatedFolder.unreadMessageCount = message.seen ? expectedUpdatedFolder.unreadMessageCount
      : expectedUpdatedFolder.unreadMessageCount - 1;
    expectedUpdatedFolder.newMessageCount = message.recent ? expectedUpdatedFolder.newMessageCount - 1
      : expectedUpdatedFolder.newMessageCount;
  });
  dispatch(updateCacheIfExist(folder, messagesToDelete, true));
  dispatch(setSelected(messagesToDelete, false));
  dispatch(updateFolder(expectedUpdatedFolder));
  dispatch(lockMessages(messagesToDelete));

  fetch(url, {
    method: 'DELETE',
    headers: credentialsHeaders(credentials)
  })
    .then(toJson)
    .then(updatedFolder => {
      dispatch(unlockMessages(messagesToDelete));
      dispatch(updateFolder(updatedFolder));
      dispatch(deleteFromCache(updatedFolder, messagesToDelete));
    })
    .catch(() => {
      dispatch(unlockMessages(messagesToDelete));
      // Rollback state from dispatched expected responses
      dispatch(updateCacheIfExist(folder, messages));
      dispatch(updateFolder(folder));
    });
}
