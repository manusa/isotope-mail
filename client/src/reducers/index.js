import {combineReducers} from 'redux';
import application from './application';
import folders from './folders';
import messages from './messages';

export const INITIAL_STATE = {
  application: {
    title: 'Isotope Mail Client',
    user: {},
    selectedFolder: {},
    activeRequests: 0
  },
  folders: {
    selected: {},
    items: [],
    activeRequests: 0
  },
  messages: {
    cache: {},
    activeRequests: 0
  }
};

// https://github.com/reduxjs/redux/issues/749#issuecomment-164327121
export default combineReducers({
  application,
  folders: folders,
  messages: messages
});
