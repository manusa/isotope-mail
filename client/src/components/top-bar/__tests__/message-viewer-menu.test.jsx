import React from 'react';
import {shallow} from 'enzyme';
import * as messageService from '../../../services/message';
import MessageViewerMenuConnected, {MessageViewerMenu} from '../message-viewer-menu';
import {createMockStore} from '../../../__testutils__/store';
import {INITIAL_STATE} from '../../../reducers';

describe('MessageViewerMenu component test suite', () => {
  let t;
  beforeAll(() => {
    t = jest.fn(arg => arg);
  });

  describe('Snapshot render', () => {
    test('Snapshot render, visible, should render message-viewer-menu', () => {
      // Given
      const props = {t, selectedFolder: {}, selectedMessage: {}, visible: true};
      // When
      const messageViewerMenu = shallow(<MessageViewerMenu {...props}/>);
      // Then
      expect(messageViewerMenu).toMatchSnapshot();
    });
    test('Snapshot render, visible=false, should render message-viewer-menu "hidden', () => {
      // Given
      const props = {t, selectedFolder: {}, selectedMessage: {}, visible: false};
      // When
      const messageViewerMenu = shallow(<MessageViewerMenu {...props}/>);
      // Then
      expect(messageViewerMenu).toMatchSnapshot();
    });
  });
  describe('Events tests', () => {
    test('Download click, should invoke downloadMessage', () => {
      // Given
      const props = {t, selectedFolder: {}, selectedMessage: {}, visible: true, downloadMessage: jest.fn()};
      const messageViewerMenu = shallow(<MessageViewerMenu {...props}/>);
      // When
      messageViewerMenu.find('li').simulate('click', {});
      // Then
      expect(props.downloadMessage).toHaveBeenCalledTimes(1);
    });
  });
  describe('Redux properties', () => {
    test('setMessageFilter, should invoke application setMessageFilterKey action', () => {
      // Given
      messageService.downloadMessage = jest.fn();
      const store = createMockStore({...INITIAL_STATE});
      const messageViewerMenu = shallow(<MessageViewerMenuConnected store={store} />);
      // When
      messageViewerMenu.props().downloadMessage();
      // Then
      expect(messageService.downloadMessage).toHaveBeenCalledTimes(1);
    });
  });
});
