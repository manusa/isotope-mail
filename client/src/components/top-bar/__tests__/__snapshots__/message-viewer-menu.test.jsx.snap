// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageViewerMenu component test suite Snapshot render Snapshot render, visible, should render message-viewer-menu 1`] = `
<div
  aria-hidden={false}
  className="message-viewer-menu mdc-menu mdc-menu-surface
      mdc-menu-surface--open"
>
  <ul
    className="mdc-list mdc-list--dense"
  >
    <li
      className="mdc-list-item"
    >
      topBar.messageViewerMenu.download
    </li>
  </ul>
</div>
`;

exports[`MessageViewerMenu component test suite Snapshot render Snapshot render, visible=false, should render message-viewer-menu "hidden 1`] = `
<div
  aria-hidden={true}
  className="message-viewer-menu mdc-menu mdc-menu-surface
      "
>
  <ul
    className="mdc-list mdc-list--dense"
  >
    <li
      className="mdc-list-item"
    >
      topBar.messageViewerMenu.download
    </li>
  </ul>
</div>
`;
